sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessorBase.cs

```diff
@@ -0,0 +1,729 @@
+                    return ClaimOwnershipAsync(unclaimedPartitions.ElementAt(index), completeOwnershipEnumerable);
+                }
+
+                // Only try to steal partitions if there are no unclaimed partitions left.  At first, only processors that have exceeded the
+                // maximum owned partition count should be targeted.
+
+                var maximumOwnedPartitionsCount = minimumOwnedPartitionsCount + 1;
+
+                IEnumerable<string> stealablePartitions = activeOwnership
+                    .Where(ownership => partitionDistribution[ownership.OwnerIdentifier] > maximumOwnedPartitionsCount)
+                    .Select(ownership => ownership.PartitionId);
+
+                // Here's the important part.  If there are no processors that have exceeded the maximum owned partition count allowed, we may
+                // need to steal from the processors that have exactly the maximum amount.  If this instance is below the minimum count, then
+                // we have no choice as we need to enforce balancing.  Otherwise, leave it as it is because the distribution wouldn't change.
+
+                if (!stealablePartitions.Any()
+                    && ownedPartitionsCount < minimumOwnedPartitionsCount)
+                {
+                    stealablePartitions = activeOwnership
+                        .Where(ownership => partitionDistribution[ownership.OwnerIdentifier] == maximumOwnedPartitionsCount)
+                        .Select(ownership => ownership.PartitionId);
+                }
+
+                // If any stealable partitions were found, randomly pick one of them to claim.
+
+                if (stealablePartitions.Any())
+                {
+                    var index = RandomNumberGenerator.Value.Next(stealablePartitions.Count());
+
+                    return ClaimOwnershipAsync(stealablePartitions.ElementAt(index), completeOwnershipEnumerable);
+                }
+            }
+
+            // No ownership has been claimed.
+
+            return Task.FromResult<PartitionOwnership>(null);
+        }
+
+        /// <summary>
+        ///   Stops an owned partition processing task in case it is running.  It is also removed from the tasks dictionary
+        ///   along with its corresponding token source.
+        /// </summary>
+        ///
+        /// <param name="partitionId">The identifier of the Event Hub partition whose processing is being stopped.</param>
+        /// <param name="reason">The reason why the processing for the specified partition is being stopped.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private async Task StopPartitionProcessingIfRunningAsync(string partitionId,
+                                                                 ProcessingStoppedReason reason)
+        {
+            // TODO: what if there is no token source?  It could happen if the user added a task by themself.
```

Comment:
If we don't have knowledge of it because a developer chose not to build on our infrastructure, we cannot be responsible for managing it.

Reply:
Yeah. I think it can't be helped.